route/route.go

```diff
@@ -626,3 +646,119 @@ func getFirstValueFromMetadata(key string, md metadata.MD) string {
 	}
 	return ""
 }
+
+type environmentCache struct {
+	mutex sync.RWMutex
+	items map[string]*cacheItem
+}
+
+func newEnvironmentCache() *environmentCache {
+	return &environmentCache{
+		items: make(map[string]*cacheItem),
+	}
+}
+
+type cacheItem struct {
+	expiresAt time.Time
+	value     string
+}
+
+func (c *environmentCache) get(key string) string {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	if item, ok := c.items[key]; ok {
+		if time.Now().Before(item.expiresAt) {
+			return item.value
+		}
+	}
+
+	return ""
+}
+
+func (c *environmentCache) getOrSet(key string, ttl time.Duration, getFn func(string) (string, error)) (string, error) {
+	val := c.get(key)
+	if val != "" {
+		return val, nil
+	}
+
+	c.mutex.Lock()
```

Comment:
we could probably move the write lock down to `addItem` function - it's the lowest level that writes, and we might call it from elsewhere

Reply:
I was aiming to let the write block happen on concurrent requests so we don't do parallel calls to auth endpoint, but on reflection it just makes them serial instead which isn't any better.

Instead, I'll do another cache check after acquiring the lock and if still no result then do the auth check. Then if multiple calls see a cache miss, only the first one to get the write lock will do the auth check and populate the cache. The following write locks will then get a cache hit and skip the auth check.

Reply:
I ended up following exactly this pattern in Basenji, so I agree with this decision.

Reply:
Thanks both 👍🏻